### PR TITLE
144 merge job and experiment slices

### PR DESCRIPTION
**Pull request by OsamaMSoliman**

### CODE REVIEW

File: src/ml_board/frontend/Dashboard/src/redux/experiments/yetAnotherExperimentSlice.tsx [hunk 1]
```diff
@@ -1,66 +1,67 @@
...
 
-export const { upsertExperiment, updateExperiment } = experimentsSlice.actions;
+// export const { upsertExperiment, updateExperiment } = experimentsSlice.actions;
 
-// create a set of memoized selectors
-export const {
-    selectById: selectExperimentById,
-    selectEntities: selectExperiments,
-} = experimentsAdapter.getSelectors((state: RootState) => state.experiments)
+// // create a set of memoized selectors
```
Comment (le1nux) [inline src/ml_board/frontend/Dashboard/src/redux/experiments/yetAnotherExperimentSlice.tsx:60]: legacy code

File: src/ml_board/frontend/Dashboard/src/redux/experiments/yetAnotherExperimentSlice.tsx [hunk 2]
```diff
@@ -1,66 +1,67 @@
...
-        //   state[action.payload.experiment_id] = { ...action.payload, color: getRandColor() };
-        // } else {
-        //   state[action.payload.experiment_id] =
-        //     mergeExperiments(state[action.payload.experiment_id], action.payload);
-        // }
-        // }
-        updateExperiment: experimentsAdapter.updateOne,
-    }
-});
+// export const experimentsSlice = createSlice({
```
Comment (le1nux) [inline src/ml_board/frontend/Dashboard/src/redux/experiments/yetAnotherExperimentSlice.tsx:31]: code that is commented out is legacy code ;-)

File: src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import { DataFromSocket, DataToRedux } from "./DataTypes";
+import { MLGYM_EVENT } from "./EventsTypes";
+import handleEvaluationResultData, { evalResultCustomData, EvaluationResultPayload } from "./event_handlers/evaluationResultDataHandler";
+import handleExperimentStatusData from "./event_handlers/ExperimentStatusHandler";
+import handleJobStatusData from "./event_handlers/JobStatusHandler";
+
+// Note: I don't like having anything here other than the callbacks but it is what it is :')
```
Comment (le1nux) [inline src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts:7]: I think it is even good practice to separate things  like this. :-)

File: src/ml_board/frontend/Dashboard/src/worker_socket/WorkerSocket.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import socketIO, { Socket } from 'socket.io-client';
```
Comment (le1nux) [inline src/ml_board/frontend/Dashboard/src/worker_socket/WorkerSocket.ts:1]: this code is very well structured. Higher level functions call lower level functions and therefore it reads like a story. Good work! :-)

File: src/ml_board/frontend/Dashboard/src/worker_socket/WorkerSocket.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+const onPongReceivedfromWebsocketServer = () => {
+    // on Pong , save time of receiving 
+    lastPong = new Date().getTime();
+    // calculate the ping and send it to the MainThread
+    pingMainThreadCallback(lastPong - lastPing);
+};
+
+// ========================= helper methods ============================//
+
+// ASK MAX: sendPingToServer or sendPing or pingToServer or pingingServer or pinging
```
Comment (le1nux) [inline src/ml_board/frontend/Dashboard/src/worker_socket/WorkerSocket.ts:67]: send_ping_to_websocket_server is perfect :+1:

**Review by le1nux — APPROVED** [review]
Looks good to me. :+1:  Before merging, please remove all the code that is commented out / that is not needed anymore.

File: src/ml_board/frontend/Dashboard/src/worker_socket/WorkerSocket.ts (hunk 4 above)
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/worker_socket/WorkerSocket.ts:1]: Thanks, that's the whole reason why I have redone it :D glad you like it

File: src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts (hunk 3 above)
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts:7]: I have a good news from a future PR, all the variables except the map are removed (not needed) ;)

File: src/ml_board/frontend/Dashboard/src/redux/experiments/yetAnotherExperimentSlice.tsx [hunk 6]
```diff
@@ -1,66 +1,67 @@
-import { createEntityAdapter, createSlice, EntityState } from "@reduxjs/toolkit";
```
[the remark this answers is not in the document]
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/redux/experiments/yetAnotherExperimentSlice.tsx:1]: Well this entire file is deleted in a next PR 😅 so no need to rename it

**Comment by OsamaMSoliman** [comment]
> Looks good to me. 👍 Before merging, please remove all the code that is commented out / that is not needed anymore.

I usually comment the code in one commit and delete it in the following one, that's in order to have it for longer if someone notices something is not working as expected or so, instead of reverting both codes can be seen next to each other!

so the commented codes (or the entire commented files) are deleted in the next PR (ideally! or maybe the one after that, I don't quite remember😅)

**Pull request merged**